File: frontend/apps/marveline/src/schemas/pricing.ts
import { z } from 'zod'
import { dateIsoSchema, idSchema } from './common'

export const pricingTierSchema = z.object({
  min_qty: z.number().int().min(1),
  max_qty: z.number().int().positive().optional(),
  unit_price_cents: z.number().int().min(0, 'Prix invalide'),
})

export const pricingRuleSchema = z
  .object({
    name: z.string().min(1, 'Nom requis'),
    rule_type: z.enum(['flat', 'per_day', 'tiered', 'volume', 'seasonal', 'custom']),
    applies_to: z.enum(['product', 'category', 'all']),
    target_id: idSchema.optional(),
    discount_pct: z.number().min(0).max(100).optional(),
    tiers: z.array(pricingTierSchema).optional(),
    valid_from: dateIsoSchema.optional(),
    valid_to: dateIsoSchema.optional(),
  })
  .refine(
    (d) => {
      if (d.valid_from && d.valid_to) {
        return new Date(d.valid_to) >= new Date(d.valid_from)
      }
      return true
    },
    { message: 'La date de fin doit être après la date de début', path: ['valid_to'] },
  )
  .refine(
    (d) => d.rule_type !== 'tiered' || (d.tiers && d.tiers.length > 0),
    { message: 'Les règles par paliers nécessitent au moins un palier', path: ['tiers'] },
  )

export type PricingRuleInput = z.input<typeof pricingRuleSchema>

File: app/models/pricing.py
"""Modèles ORM pour les règles de pricing."""
from datetime import date
from typing import Optional, List
from sqlalchemy import BigInteger, Boolean, Date, ForeignKey, Integer, String
from sqlalchemy.orm import Mapped, mapped_column, relationship

from app.models.base import Base, TimestampMixin, TenantMixin


class PricingRule(Base, TimestampMixin, TenantMixin):
    """Règle de tarification (flat, par jour, volume, saisonnier…)."""

    __tablename__ = "pricing_rules"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    name: Mapped[str] = mapped_column(String(255), nullable=False)
    rule_type: Mapped[str] = mapped_column(String(30), nullable=False)
    # flat | per_day | tiered | volume | seasonal | custom
    applies_to: Mapped[str] = mapped_column(String(20), nullable=False)
    # product | category | all
    target_id: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    discount_pct: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    valid_from: Mapped[Optional[date]] = mapped_column(Date, nullable=True)
    valid_to: Mapped[Optional[date]] = mapped_column(Date, nullable=True)
    active: Mapped[bool] = mapped_column(Boolean, nullable=False, default=True)

    tiers: Mapped[List["PricingTier"]] = relationship(
        "PricingTier", back_populates="rule", cascade="all, delete-orphan"
    )

    def __repr__(self) -> str:
        return f"<PricingRule id={self.id} name={self.name!r} type={self.rule_type}>"


class PricingTier(Base, TenantMixin):
    """Palier de prix associé à une règle tiered/volume."""

    __tablename__ = "pricing_tiers"

    id: Mapped[int] = mapped_column(Integer, primary_key=True, autoincrement=True)
    rule_id: Mapped[int] = mapped_column(
        Integer, ForeignKey("pricing_rules.id", ondelete="CASCADE"), nullable=False
    )
    min_qty: Mapped[int] = mapped_column(Integer, nullable=False)
    max_qty: Mapped[Optional[int]] = mapped_column(Integer, nullable=True)
    unit_price_cents: Mapped[int] = mapped_column(BigInteger, nullable=False)

    rule: Mapped["PricingRule"] = relationship("PricingRule", back_populates="tiers")

    def __repr__(self) -> str:
        return f"<PricingTier id={self.id} rule_id={self.rule_id} min_qty={self.min_qty}>"
